File: packages/agent-sdk-python/src/agentmarket_sdk/cost_estimator.py
"""Cost estimation — the Python port of `packages/agent-sdk/src/cost-estimator.ts`."""

from __future__ import annotations

import httpx

from .http import build_url, read_payment_required
from .types import ATOMIC_UNITS_PER_USD, CallParams, CostEstimate


async def estimate_cost(
    http_client: httpx.AsyncClient,
    base_url: str,
    resource: str,
    params: CallParams | None = None,
) -> CostEstimate:
    """Prices a resource without paying for it — x402 already hands this to
    us for free: a request with no X-PAYMENT header gets a 402 back with
    the exact price, and the server never charges (or even logs a payment
    attempt) for that response. No separate pricing endpoint needed."""
    res = await http_client.get(build_url(base_url, resource, params))

    if res.status_code != 402:
        raise RuntimeError(
            f'Expected 402 Payment Required while estimating the cost of "{resource}", '
            f"got HTTP {res.status_code} instead — is this a metered endpoint?"
        )

    body = read_payment_required(res)
    if not body.accepts:
        raise RuntimeError(f'Server returned a 402 for "{resource}" with no payment options in "accepts".')

    requirement = body.accepts[0]
    atomic = float(requirement.amount if requirement.amount is not None else requirement.max_amount_required)
    return CostEstimate(
        resource=resource,
        price_usd=atomic / ATOMIC_UNITS_PER_USD,
        network=requirement.network,
        scheme=requirement.scheme,
    )

File: packages/agent-sdk/src/cost-estimator.ts
import { buildUrl, readPaymentRequired } from './http.js';
import { ATOMIC_UNITS_PER_USD, type CallParams, type CostEstimate } from './types.js';

/**
 * Prices a resource without paying for it — x402 already hands this to us
 * for free: a request with no X-PAYMENT header gets a 402 back with the
 * exact price, and the server never charges (or even logs a payment
 * attempt) for that response. No separate pricing endpoint needed.
 */
export async function estimateCost(
  fetchImpl: typeof fetch,
  baseUrl: string,
  resource: string,
  params?: CallParams,
): Promise<CostEstimate> {
  const res = await fetchImpl(buildUrl(baseUrl, resource, params));

  if (res.status !== 402) {
    throw new Error(
      `Expected 402 Payment Required while estimating the cost of "${resource}", got HTTP ${res.status} instead — is this a metered endpoint?`,
    );
  }

  const body = await readPaymentRequired(res);
  const requirement = body.accepts[0];
  if (!requirement) {
    throw new Error(`Server returned a 402 for "${resource}" with no payment options in "accepts".`);
  }

  const atomic = Number(requirement.amount ?? requirement.maxAmountRequired);
  return {
    resource,
    priceUsd: atomic / ATOMIC_UNITS_PER_USD,
    network: requirement.network,
    scheme: requirement.scheme,
  };
}
